Use Debouncer for update calls on stats sensors

azogue opened the pull request:
A fresh restart of #32088, using **helpers/debounce** (suggested by @balloob)
to limit activity when incoming bursts of samples are received.

For the debouncer: set cooldown=10ms and immediate=False (we want to wait for the sampling burst to end before calling the sensor update)

cc @exxamalte @MartinHjelmare @fabaff 

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Instead of call `await self.async_update_ha_state(True)` after each sample and also inside the purge callback, we call a debounced wrap of that: `self.hass.async_create_task(self._debounced_updater.async_call())`

That way we ensure that multiple simultaneous calls are processed just once, and the number of cancelling+setting the purge callback is minimized, thus decreasing the stress on the event bus.

As the issue (#31997) is hard to catch, this fix may need a test _in real conditions_ to be sure that it is valid, and the `unknown listener` core warnings completely disappear. I will put it in my _production HA_ to check it this week, and I'll feedback with the results.

#### Notes: `Debouncer` usage and tests

cc @balloob 

When debouncer is set to immediate=False, and the call is not done until _cooldown_ is expired, the common `hass.block_till_done()` does not wait for the debouncer (there are no tasks registered), so `async_update` is never called.

To _quickfix_ that I introduced a silly task (a call to `await asyncio.sleep(cooldown)`) to prevent the block from finishing before the delayed update is called. 

That may require a deeper review to solve it with a more elegant hack :)
The issue only affects tests, as _in real time_ things happen when they need to.




## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes  #31997
- This PR is related to issue:  #31576
- This PR is another approach of #32088, so at least one of the 2 should be closed.
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
Use Debouncer for update calls on stats sensors

to limit activity when incoming bursts of samples are received.

Set cooldown=10ms and immediate=False (we want to wait for the burst
to end before calling sensor update)

balloob commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -293,7 +318,7 @@ def _next_to_purge_timestamp(self):
                 """Timer callback for sensor update."""
                 _LOGGER.debug("%s: executing scheduled update", self.entity_id)
-                self.async_schedule_update_ha_state(True)
+                self.hass.async_create_task(self._debounced_updater.async_call())
```
You should await it. But do that after you set `self._update_listener = None`

azogue replied:
Done for both

balloob commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -112,8 +138,7 @@ def async_stats_sensor_state_listener(entity, old_state, new_state):
             self._add_state_to_queue(new_state)
-
-            self.async_schedule_update_ha_state(True)
+            self.hass.async_create_task(self._debounced_updater.async_call())
```
just await it directly. No need to create a task. Requires you to change `@callback` -> `async def`

azogue replied:
Done

balloob commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -100,10 +115,21 @@ def __init__(self, entity_id, name, sampling_size, max_age, precision):
     async def async_added_to_hass(self):
         """Register callbacks."""
 
+        async def _async_debounced_update():
```
Why only set this in `async_added_to_hass` ? You can do this in the constructor.

azogue replied:
I was playing with `self.hass._` at start (which I only have access in added, I think), but yes, I'll move it. 

Another doubt I had was where to instance the Debouncer, in setup or in entity init. It's ok in setup, right?

balloob commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -71,9 +72,15 @@
     sampling_size = config.get(CONF_SAMPLING_SIZE)
     max_age = config.get(CONF_MAX_AGE, None)
     precision = config.get(CONF_PRECISION)
+    debounced_updater = Debouncer(hass, _LOGGER, cooldown=0.01, immediate=False)
```
Why `immediate` not set to `True` ?

azogue replied:
because if 3 samples are received, we want to make the update (data process) with all samples, so at the end of the grouped calls

azogue replied:
(If I understood correctly the code, both in debouncer and in statistics)

Update call does the purging, process of all attributes, and state definition. If called in the first sample, the data processed is not correct, as does not mirror all the received samples, right?

balloob replied:
I just realized that I've updated our debouncer like a throttle. It will ensure a maximum of calling a command once per `<cooldown>` period, instead of after `<cooldown>` period, with each call resetting the `<cooldown>`.

balloob replied:
I also realized that it isn't 100% perfect, so added some fixes in #32161

azogue committed:
Apply review changes

azogue commented:
> As the issue (#31997) is hard to catch, this fix may need a test in real conditions to be sure that it is valid, and the unknown listener core warnings completely disappear. I will put it in my production HA to check it this week, and I'll feedback with the results.

**---> This patch does not solve the problem :(**

```log
2020-02-25 02:11:11 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe98bb198>
2020-02-25 02:12:10 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe9a27660>
2020-02-25 02:13:11 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe98ae8a0>
2020-02-25 03:14:42 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe9840f60>
2020-02-25 03:16:42 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe7ac5ae0>
2020-02-25 03:18:12 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xeb0908a0>
2020-02-25 05:56:15 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xeb0dbd20>
2020-02-25 06:27:09 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe9861780>
2020-02-25 06:28:41 WARNING (MainThread) [homeassistant.core] Unable to remove unknown listener <function async_track_point_in_utc_time.<locals>.point_in_time_listener at 0xe793fe40>
```

They are there AGAIN :(

It could be because of the short cooldown (just 10ms), or because the lack of lock (#32161), but I'm not sure right now :(

The patch in #32088, with the naive approach of filter some purge schedule, did the trick. I will test some more, as I see more interesting to work with the debouncer

azogue commented:
Maybe this is the key to, at least, remove the core warnings:

```python
if self._update_listener:
    if not hasattr(self._update_listener, "run"):
        self._update_listener()
    else:
        _LOGGER.warning("%s: Gotcha, this was going to generate a core warning :)", self.entity_id)
    self._update_listener = None
```
(print is always the debugger best-friend)

azogue commented:
cc @ballob, @exxamalte 

> And the thing is that with 3 lines of code to wrap the update and call the debouncer instead, 
> the issue disappears completely, as now the update calls are grouped when these bursts come in.

Sadly, that is not true, only reduces the frequency of warnings raising :(

But I think I've got it, finally.

### What is happening?

After a lot of approaches,  I've identified **multiple causes** for the extra stress that, apparently, the stats sensor is creating in the event bus with the max age purge scheduling.

1. **Burst incoming**: sometimes a burst of new samples is received, and for each one the cycle of `purge + update + cancel old listener + create new one` is repeated.
2. **Schedule repeat**: Most times the new schedule matches the last one, or it is very near it, but this is not checked, and instead we cancel + reprogram it.
3. **Periodic overlapping**: Most input signals for the stats sensor will have a constant frequency, with period T, and with a `max_age = n * T`, the scheduled purge of the oldest sample will be called **~almost at the same time** than the next incoming sample
  
All of these are happening in **sub-second intervals**, while the event bus fires `EVENT_TIME_CHANGED` only once each second, and only fires `EVENT_TIMER_OUT_OF_SYNC` if the event is > 1 sec late, so there is a window of max 1 sec where a tick can be a little late and we don't know.

So, this is the only thing I can imagine that it is happening for the issue to appear:

- current purge schedule is waiting in 'listener X' set to t=-.7, so tick in t=-1 does not activate it,
- a tick with t=0 is fired, and matched 'listener X' gets added for evaluation
- 'listener X' is called with tick t=0, gets "run" tag, **does the __unsub__**, and adds the 'purge job X' to queue,
- **new sample comes in t=0.5, calling update. As `_update_listener` is not None, it is canceled, but because it is not there, the warning is generated**.
- 'purge job X' callback is finally called, setting `_update_listener` to None too late.

This is only my hypothesis, as I've been incapable of reproducing the issue upon request, and every approach worked ok in the dev machine. The __'good thing'__ is that I'm suffering of this core warning in the home machine, with different frequencies for each approach, so I've been __real testing__ all approaches in that environment (a RPI4_4GB+USB_SSD running hassio 105.5 on docker, editing the files inside the HA container and restarting it)

Along the way I've discovered/validated some things:

- All warnings are coming from the stats sensors with max age, for sure.
- All unknown listeners have the "run" tag added, so they've been called (or **just added the job to queue**) in the past, and because of `async_unsub()`, they have removed the listener.
- The timestamp of the lost listeners has ~always a datetime of <1s near the core warning 

And I've also made a lot of small changes to try to remove the issue. Finally, I have a working version running for > 14 h without the issue, so I hope it is fixed, with these:

### Changes / Optimizations 

- **Debounced call** to `async_update` to handle bursts of incoming samples.
- Abort call to update from an incoming sample if a scheduled purge is coming in near future.
- Store the last scheduled purge and compare times with new one, to abort re-scheduling to the same (or very similar) future timestamp. Most important optimization by far.
- Access to `self._update_listener` with a lock around (I'm almost sure that this is unnecessary, but current working version has it)
- **Do not operate in sub-second times**: as the HA ticking is of 1 second, (scheduling listeners to HH:MM:SS.6 makes the listener run the callback in 1.0 anyway).
  Instead, admit up to 1 second error in the max_age of the sensor (which is very reasonable IMHO), by __ceiling__ the purge scheduled timestamps. This one is where the magic happens :)

#### Results

To put some numbers on these, here is the evolution of the RPI machine for the last hours:

![image](https://user-images.githubusercontent.com/9748301/75433933-0cc9db80-5951-11ea-806b-c4a84a5d7045.png)

- It shows that ~1/3 of the updates are scheduled purge operations and 2/3 are incoming samples
- Yellow area is the number of aborted cancel + re-schedule actions (before the fix those where happening)
- Small in comparison, but also important, green area shows the aborted calls to update from incoming samples
- The debouncing doesn't appear here, but it's also happening (I did not introduce logging messages in debouncer, so I have no data, but in dev machine I've seen it working with groups of 2 and 3 calls)

Plot notes:
* Data comes from `/api/error_log` after setting stats at debug level, parsing the generated log messages. So far no warnings.
* There are 4 stats sensors with max_age from 5 to 15 min, all of them filtering template sensors dependant of 2 to 6 entities each one.
* It plots a `resample(2min).count()`, so a value of 120 means one event per sec.

Logging notes:
* I introduced a small hack in `async_track_point_in_utc_time` to tag the listeners, with:

  ```python
  # Set variable to identify origin (for error logging purposes).
  setattr(
      point_in_time_listener,
      "__name__",
      f"{action.__name__}_{point_in_time:%H_%M_%S_%f}",
  )
  ```
  
  That way I could see the issues like that:
  
  ```log
  2020-02-26 12:49:41.000 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 113) named '_scheduled_update_11_49_40_000000' (run True)
  2020-02-26 13:19:03.025 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 118) named '_scheduled_update_12_19_02_648276' (run True)
  2020-02-26 14:03:41.069 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 114) named '_scheduled_update_13_03_40_422252' (run True)
  2020-02-26 14:05:41.047 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 116) named '_scheduled_update_13_05_40_342420' (run True)
  2020-02-26 14:09:04.028 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 114) named '_scheduled_update_13_09_03_581080' (run True)
  2020-02-26 14:12:04.112 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 115) named '_scheduled_update_13_12_03_613611' (run True)
  2020-02-26 14:14:34.025 ERROR (MainThread) [homeassistant.core] Unable to remove unknown 'time_changed' listener (there are 116) named '_scheduled_update_13_14_33_059348' (run True)
  ```

I'll be pushing the last changes here for review, after rebasing with last changes in debouncer, 
and I will update the RPI to 106 to check the original issue is there, and then again with the described patch (maybe without the extra lock, to check that it is not necessary).

azogue committed:
Add delta error for max_age to reduce purge scheduling

- Use a max_age_error of 1 second, and ceil purge schedules
  to exact seconds, removing sub-second mess.
- Compare with next expected purge before calling update
  on incoming sample, to abort the call if a new update comes
  inside the error margin.
- Compare with last scheduled before canceling last listener, to abort
  re-scheduling if the last one is the same or it's inside the margin
- Use a lock to check `self._update_listener` (may be not necessary)
- In startup from recorder, add a little extra for the initial purge,
  so the initial updates don't raise the warning with the initial stress
  in event bus.

azogue committed:
Remove unneeded lock and fix some log messages

azogue committed:
Remove todo

azogue committed:
Add wait_until_cooldown to debouncer

so block_til_done waits for delayed debouncers (immediate=False)

azogue commented:
> I'll be pushing the last changes here for review, after rebasing with last changes in debouncer,

Done

> and I will update the RPI to 106 to check the original issue is there, and then again with the described patch (maybe without the extra lock, to check that it is not necessary).

Also done.
- Original issue still happens in 106.0 (no surprises)
- This fix also works on 106, without extra locks, so I removed it.

@balloob please check https://github.com/home-assistant/home-assistant/pull/32155/commits/476a165bfadbec571f09cb39feae77cd0631f3b6, about the delayed debouncer + tests using `hass.block_til_done()`

balloob commented on homeassistant/helpers/debounce.py:
```diff
@@ -104,3 +104,7 @@ def _schedule_timer(self) -> None:
             self.cooldown,
             lambda: self.hass.async_create_task(self._handle_timer_finish()),
         )
+        self.hass.async_create_task(self._wait_until_cooldown())
```
I don't see what value this adds, you're just scheduling a task and no one is waiting for it ?

azogue replied:
Yeah, basically, and is not pretty.

This is what I was trying to tell you in the debouncer PR: tests rely on `block_til_done` to wait for all pendent tasks to finish, but debouncers set to immediate=False won't be awaited, as there are no tasks registered.

t=0 -> A new sample is received, debouncer starts cooldown (but no call to update)
t=0.001 -> Block til done is called to wait for the sample to be processed, but as there are no tasks, it returns instantly.
t=0.05 -> Cool down finishes and update is finally called (but it is too late for the test)

So I did that **for the statistics sensor tests to pass**, but I don't like it. I suppose I could insert artificial waits inside the tests, to make them pass, but I thought that option would be even uglier, not sure now. 

I was expecting to discuss how to approach this problem :) (which is only for tests, as in real-time debouncers end ok, and block_til_done() is not called)

azogue replied:
> I could insert artificial waits inside the tests, to make them pass, but I thought that option would be even uglier, not sure now.

**I reverted the change in debouncer and inserted the wait in the tests**, but the problem I see is that anyone using `hass.block_till_done()`, without specific knowledge of how the debouncer works, is going to expect that the call is made, so it could be a little confusing, right?

balloob replied:
We should not sleep in tests. That is slowing down our tests. Instead catch the call to schedule it, and run that directly.

azogue replied:
Ok, I'll make the change

balloob commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -99,26 +122,48 @@ def __init__(self, entity_id, name, sampling_size, max_age, precision):
 
-            self.async_schedule_update_ha_state(True)
+            if self._max_age:
+                # Do not call update if a purge update is scheduled in the short time
```
This should not be necessary? 

Purging is only done inside `async_update`, which is only being called via the debouncer. The debouncer is smart enough that if the call is running, it will cancel the call too (after my latest fix). So you shouldn't need this logic.

azogue replied:
Hi @balloob, thanks for reviewing :)

The debouncer operates in a time window of **50 ms**, and addresses 1 of the 3 problems detected: the burst incoming, but for the other 2 (schedule repeat and periodic overlapping) I'm admitting up to **1 second** error in the max_age of the sensor. These are very different time windows.

If you see the plot with the new behavior, the number of aborted calls to update from incoming samples (the small green area) is small compared to the other aborts, but I'm not sure if removing that the issue will appear again.

After all the tests I did, I think the key to making it work was to **avoid doing sub-second operations** (as the scheduler uses a 1-second ticking clock). On that assumption, if a sample is received in t=0, but next purging is scheduled in t=0.8, the new sample is added BUT update is not called, as it will be in less than a second.

balloob replied:
But I don't understand, calling the debouncer actually doesn't abort anything because it's actually behaving like a throttle and not a debouncer (I should change the name in another PR).

azogue replied:
Sorry @balloob, I was AFK for some days...

> calling the debouncer actually doesn't abort anything because it's actually behaving like a throttle and not a debouncer

I suppose you mean that, with a cooldown of 50ms, given a really long burst of samples, like a lot of samples over 500ms:
*  a real debouncer would do only a final call, at the end of the burst.
*  and the throttle would do ~10 calls, once each 50 ms
right?

(I ask you this because I'm not sure if I'm kind of lost in translation, as English is not my best skill :)

If we are in sync, I'll try to explain myself a little better :)

- The bursts I detected when doing the patch were very close in time: usually 2 or 3 samples in a window of < 25 ms. I didn't detect the problem of a continued burst over time, so for that case, there is no difference between the throttle and the debouncer, as both would do only one call, at the end of the cooldown time. So the current (in _dev_) implementation works as expected to solve this particular problem.
  
  In that context,
  > calling the debouncer actually doesn't abort anything
  
  When a burst of three samples comes, the three samples are added, but only the third one calls the update, as it's desired to happen, so I don't follow you there :(

- But there are more problems caused by incoming samples or schedule updates happening in the **sub-second** window (but separated in time with > 5 * cooldown).

- Imposing a debouncer _aka throttler_ **with a cooldown of 1 full second** would probably fusion both kind of problems, without any timestamp logic to decide calls, but as the debouncer is in `immediate=False`, that would impose a much bigger delay for the sensor to update, and I thought 1-second delay is too much and is more a new bug than a fix :)

- That's why the debouncer handles the burst problem, with a very narrow time-window of 50ms, and the sub-second problem is removed with these **rules of "wait for the next scheduled update" or "do not re-schedule the ~same timestamp"**.


Having said all that, I also think that maybe the original problem could be solved in a simplest way, as this one is the result of multiple additions in the quest of removing the core warning. I'm only confident that this particular sum of patches is working ok.

My bet is that the debouncer could be removed from here, and just deal with the timestamp logic, and not the opposite, but I may be wrong. 

I'll try to remove that condition and do the test, to see if it does not raise the core warning.

balloob replied:
> and the throttle would do ~10 calls, once each 50 ms
right?

This is correct.

And waiting 50ms is also totally fine. 

So if you had a `helper.Debounce(cooldown=50ms, immediate=True)`, you can get this behavior with _just_ the debouncer:

T: 0ms
Sample comes in.
State is updated.

T: 10ms
Sample comes in.
Debouncer is set to update when cooldown is expired.

T: 20ms
Sample comes in

T: 50ms
Cooldown expires
State is updated.
New cooldown is set.

T: 60ms
Sample comes in
Debouncer is set to update when cooldown is expired.

balloob replied:
Which seems like desired behavior.

azogue replied:
> Which seems like desired behavior.

Not exactly :), as in your flow example, there are 2 state updates (I would want just one).

For sure that works for a continued flow like the exposed one, but as I've experienced with this sensor, there is no sample at 60ms: bursts come in small packets, with much longer intervals between them.

My proposal was to remove the initial one, following the main objective of decreasing both:
1. The excessive event bus re-scheduling cycle
2. The excessive number of state updates

The core warning is because (1), but (2) is also related. 
Anyway, maybe addressing (2) is out of context here.

I can try the behavior without this conditional (always calling the debouncer with incoming sample), and also switching to immediate=True, if you prefer.

I'll push the changes if everything works ok (and I'll remove the debouncer mocking in tests, as it is not necessary with immediate=True)

balloob replied:
The thing is that you are only thinking of your use case. For a lot of people, there will only be 1 state change, no bursts. In that case you don't want to slow it down for them. It's also fine to just increase the cooldown to 1 second. That way we just say that we update max once per second, which seems reasonable.

azogue replied:
> The thing is that you are only thinking of your use case

You're probably right, but it wasn't my intention (in fact, I had a solution for this issue for myself, by setting a fixed scan_interval for these sensors, but I wanted to solve the generic case)

I thought of the debouncer with a small cooldown to impose a small delay, prioritizing too much the bursts, as I was convinced at first that they were the origin of the issue. it seems that I was wrong.

>  It's also fine to just increase the cooldown to 1 second. That way we just say that we update max once per second, which seems reasonable

I did the change and tested it: sadly, **the warnings appear again** :(
(debouncer with 1s, immediate=True, removed this filter of 'Do not call update if a purge update is scheduled in the short time', so calling debounce update for all incoming samples)

As usual, all combinations worked without warnings in the dev machine, and all** of them failed in the RPI)

** Not all: the combination pushed here worked without warnings for almost a week. Now I'm trying a variant **WITHOUT debouncer**, just avoiding listener cancellation if it is scheduled _in the past_. --> It works too, so I'll push that here

azogue committed:
Revert wait_until_cooldown in debouncer

Instead, fix the statistics tests with an artificial cooldown wait
before calling hass.block_till_done()

azogue committed:
Mock debouncer call to avoid real delays in tests

azogue committed:
Remove debouncer and revert associated changes

Leave the max_age error, with 0.5s value, to reduce number of schedules,
and avoid the 'unknown listener' warning just by not calling
`self._update_listener()` if it was scheduled in the past.

azogue commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -209,92 +227,105 @@ def _purge_old(self):
+                self._update_listener is not None
+                and self._last_scheduled_purge > dt_util.utcnow()
+            ):
+                # cancel scheduled update only if it is in the future
```
cc @balloob 

This is the key to avoid the core warning.

The usual listener cancellation involves a listener set for some time in the _far_ future, so its cancellation and new schedule work always ok.

But, _sometimes_, the listener is set _in the past_ (:= it's already activated, or it's in the process to be).

In those cases, again _sometimes_, the async_update checks `self._update_listener` and sees it active, as `_scheduled_update` has not run yet.

In `helpers/event.py:async_track_point_in_utc_time`, this is happening:

```python
# This happened:
setattr(point_in_time_listener, "run", True)
async_unsub()  # so the listener is already removed

# But the action has not been called yet, it's in queue:
hass.async_run_job(action, now)
```

So when we call `self._update_listener()`, the core warning appears.

I was trying to solve the root causes, assuming that a lower number of updates could do the trick, but in the end, it is only to handle that particular moment in which the listener is caught in the middle.

I cannot assume that my particular use case can be extrapolated for anyone suffering #31997, but I think this plot is quite significant:

<img width="450" alt="Unknown" src="https://user-images.githubusercontent.com/9748301/75971304-e0263e80-5ed1-11ea-9842-ce8725d61688.png">
(sorry for the messy styling)


That shows the scheduled update cancel operations, where _y:=ts_sched - now (s)_. Almost all canceled listeners are scheduled in ~500 seconds in the future, but there are some points where ∆T is ~0 (negative value around ~5-50 ms, but I've seen up to -2.0 sec).

**Around 9 out of 10 of those events generate the core warning about the unknown listener**.
Filtering with `and self._last_scheduled_purge > dt_util.utcnow()` removes them (maybe coulb be necessary a little extra margin `dt_util.utcnow() + C`)

balloob replied:
You're right about `async_track_point_in_utc_time`. But the debouncer is guarding for it being called more than once. As long as the refresh hasn't executed, it won't schedule a new timer with the changes from #32161: https://github.com/azogue/home-assistant/blob/ec7c13300f02c266ffde1d5ad2e8ca3793b0cc14/homeassistant/helpers/debounce.py#L40-L44

azogue replied:
> But the debouncer is guarding for it being called more than once. As long as the refresh hasn't executed, it won't schedule a new timer

Yes, but the issue happens anyway (I tried that combination).

* With the changes in #32161 and debouncer set with 1s, it is true that the scheduling rate is maxed to 1 Hz, as the updater only runs once each second, tops.
* BUT, **the collision is between an update going on** (checking `if self._update_listener is None` in the way), **and an _unbounced callback_ scheduled a long time before**. Inside the callback, the call to update is debounced, but `async_track_point_in_utc_time` is not calling that directly.
* I suppose we could wrap it so the debouncer function IS the current callback, but anyway, we would be in the same position, as the point is that some listener can get caught in the middle between its unsub and the task running.

In the end, with simple logic to forget about listeners that may be running at the same moment, the issue is fixed, without any debouncer help

balloob replied:
The debouncer has a lock to prevent from breaking when being called when a run is in progress. 

And yes, there should be no unbounced callback. It should all go via the debouncer instance for it to work.

azogue replied:
> The debouncer has a lock to prevent from breaking when being called when a run is in progress.

That I now, and it works as expected.

> And yes, there should be no unbounced callback. It should all go via the debouncer instance for it to work.

It was also my thought. But reality always kicks in :(

I think I do not entirely follow you, as IMHO if there is a simple solution without debouncer, that should be preferable over a more complex one.

But, anyway, I'm back with the debouncer to check the behavior with debouncing the callback, and **it does not work**. In fact, it works much worse than before:
- Even if the _callback_ is debounced, the problem resides in the time between `async_unsub()` and the now debounced _action_ inside `hass.async_run_job(action, now)` (in the `async_track_point_in_utc_time` logic).
- Also, this does not avoid the fact that collisions are going to happen between incoming samples and scheduled operations, so canceling listeners needs to be done with care (:= by looking to their timestamps).
- With these changes (cooldown 1s, immediate=True, no timestamp control, callback IS debounced), everything breaks quickly, even in the dev machine. I got the core warning in debug with that:

```log
10:40:01.002 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:00.006207+00:00; #35
10:40:01.003 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#36)
10:40:01.003 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:01.003204+00:00(0:02:00)
10:40:02.007 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:03.002585+00:00; #36
10:40:02.008 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#37)
10:40:02.008 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:02.008718+00:00(0:02:00)
10:40:04.004 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:04.003315+00:00 and last_sched=2020-03-06 09:40:03.002585+00:00; #37
10:40:04.005 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#38)
10:40:04.005 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:04.005818+00:00(0:02:00)
10:40:05.009 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:04.003315+00:00 and last_sched=2020-03-06 09:40:05.003416+00:00; #38
10:40:05.010 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#39)
10:40:05.010 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:05.010232+00:00(0:02:00)
10:40:07.005 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#40)
10:40:07.006 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:07.006028+00:00(0:02:00)
10:40:08.009 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#41)
10:40:08.009 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: purging records older then 2020-03-06 09:38:08.009793+00:00(0:02:00)
```
And just inside that last update nº41, the core warning appeared, trying to cancel the current listener, set to 2020-03-06 09:40:08.003434+00:00 !!, but this time I got in in debug :)

What happened there? Let's see:
- At 09:40:08 UTC, 2 things happened:
  * A new sample is received
  * A scheduled update is called
- Murphy's Law makes the incoming sample be first then the scheduled call, so the debouncer runs that one.
- Meanwhile, `async_track_point_in_utc_time` has activated the listener, done the unsub, and put the task in the queue, but that one is not going to run because of the debouncer (and that is the one that removes the handler with `self._update_listener = None`.
- When the update operation running at 09:40:08 UTC (nº41) tries to make a new schedule, sees that there is one already, and cancels it, **finally raising the core warning** about the unknown listener.
- In fact, with this configuration, the warning appears A LOT, as most of the scheduled operations (that turn off the listener flag) are aborted with the debouncer.
- **There are more scary things in the logs above**. Looks like some listeners are being called **more than once** (yes, I know the 'run' flag should prevent that, that is why I said _scary_):
  * `async_track_point_in_utc_time` pass the `now` arg to action (I had to add `*args` to the debouncer to make this combination work)
  * I'm using this to know inside the updater who's the caller (incoming sample does not add 'now' argument)
  * As I was going a little crazy :), I introduced an internal counter in the sensor, to track the calls (the _#i_ in logs)
  * Also, I included ms in log timestamps for richer detail (they are in local time, so 10:40 is 9:40 UTC)

  At 09:40:01 UTC, it seems like that that _tick_ activates 2 listeners, as update is called twice with the same 'now' timestamp (but they run separated with the 1s cooldown):
  
  ```log
  10:40:01.002 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:00.006207+00:00; #35
  10:40:01.003 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#36)
  10:40:02.007 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update with now=2020-03-06 09:40:01.001008+00:00 and last_sched=2020-03-06 09:40:03.002585+00:00; #36
  10:40:02.008 DEBUG (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: updating statistics (#37)
  ```
  
  The second one has a different 'last_sched' date, 09:40:03 UTC, as it refers to the new schedule created in the first call nº36, which is in the future.

So in conclusion, as I see it:
- there is no good way to solve this issue with a debouncer/throttler with a window of ~1s, using `async_track_point_in_utc_time` for the purge scheduling.
- A debouncer with a much shorter cooldown is a very good tool for burst control, which is also happening here sometimes, but that is not the root cause for the issue to solve here.
- With simple timestamp logic to avoid excessive listener re-scheduling and collision, the issue totally disappears, and as a collateral benefit, the stress in the event bus is reduced (I noted around 1-2% CPU usage reduction in the RPI, from 11-12 to 9-10 %, but that is hard to prove or even be sure about)

azogue replied:
Code modifications for the analysis above are:

```python
# Debouncer config
_DEBOUNCING_COOLDOWN = 1.0
debounced_updater = Debouncer(
    hass, _LOGGER, cooldown=_DEBOUNCING_COOLDOWN, immediate=True
)
self._debounced_updater.function = self._async_debounced_update

# debounced function
async def _async_debounced_update(self, *args):
    """
    Debounced call to update entity.

    To process stats just once (in the end) when a burst of samples is received.
    """
    if args:
        # The call is coming from a scheduled update operation
        _LOGGER.warning(
            "%s: executing scheduled update with now=%s and last_sched=%s; #%d",
            self.entity_id,
            args[0],
            self._last_scheduled_purge,
            self._debug_update_counter,
        )
        self._update_listener = None
    await self.async_update_ha_state(True)

# Incoming samples:
self._add_state_to_queue(new_state)
await self._debounced_updater.async_call()

# Update method
async def async_update(self):
    """Get the latest data and updates the states."""
    self._debug_update_counter += 1
    _LOGGER.debug("%s: updating statistics (#%d)", self.entity_id, self._debug_update_counter)

    if self._max_age is not None:
        self._purge_old()

    self.count = len(self.states)

    if not self.is_binary:
        self._process_statistics()

    next_to_purge_timestamp = self._next_to_purge_timestamp()
    if next_to_purge_timestamp:
        if self._update_listener is not None:
            self._update_listener()
            self._update_listener = None

        self._last_scheduled_purge = next_to_purge_timestamp

        self._update_listener = async_track_point_in_utc_time(
            self.hass, self._debounced_updater.async_call, next_to_purge_timestamp
        )

# In debounce.py I had to add *args everywhere to make it work with async_track_point_in_utc_time
async def async_call(self, *args: Any) -> None:
   # ...
```

I will not push this, as that is an unknown listener warning factory :)

azogue replied:
_Outside the context of this PR/issue, as it is not related with the issue, some thoughts, @balloob:_

One of the things this PR evolution has convinced me is that **`async_track_point_in_utc_time` needs some change** because it is being used wrong here, IMO, and probably in more places.
- As the _clock tick_ is of 1Hz, it has no sense to program a listener in _t=1.2 s_, as the listener is going to run in _t=1.00x s_ or _t=2.00x s_ (probably the latter).
- In that context, all calls to `async_track_point_in_utc_time` should be done with exact seconds, always minimizing the delay between the expected time and the real one (activation is done when `now >= point_in_time`).
- Quick solution would be to **round** or **ceil** the given `point_in_time` inside `async_track_point_in_utc_time`, to discard the sub-second precission, but deciding between round and ceil is not clear.
